src/engine/app.rs

```diff
@@ -57,25 +57,26 @@ impl Application {
         self.states.start(self.context.lock().unwrap().deref_mut());
     }
 
-    /// Advances the game world by one tick.
-    fn advance_frame(&mut self) {
-        let events = self.context.lock().unwrap().poll_engine_events();
+    fn update_states(states: &mut StateMachine, context: &mut Context) {
```

Comment:
Maybe it would be better to do it in a single function like this:
```rust
fn advance_frame(&mut self) {
    {
        let mut ctx = self.context.lock().unwrap();
        let events = ctx.poll_engine_events();
        for e in events {
            ctx.broadcaster.publish().with::<EngineEvent>(e);
        }

        let entities = ctx.broadcaster.poll();
        self.states.handle_events(&entities, ctx.deref_mut());

        let fixed_step = ctx.fixed_step;
        let last_fixed_update = ctx.last_fixed_update;

        if last_fixed_update.elapsed() >= fixed_step {
            self.states.fixed_update(ctx.deref_mut());
            ctx.last_fixed_update += fixed_step;
        }

        self.states.update(ctx.deref_mut());
    }
    self.states.run_processors(self.context.clone());
    {
        let mut ctx = self.context.lock().unwrap();
        ctx.broadcaster.clean();
        ctx.renderer.submit();
    }
}
```